Library/Homebrew/rubocops/components_order.rb

```diff
@@ -49,50 +51,113 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
+            on_linux_blocks = find_blocks(resource_block.body, :on_linux)
+
+            if on_macos_blocks.length.zero? && on_linux_blocks.length.zero?
+              # Found nothing. Try without .body as depending on the code,
+              # on_macos or on_linux might be in .body or not ...
+              on_macos_blocks = find_blocks(resource_block, :on_macos)
+              on_linux_blocks = find_blocks(resource_block, :on_linux)
+
+              next if on_macos_blocks.length.zero? && on_linux_blocks.length.zero?
+            end
+
+            @offensive_node = resource_block
+            @offense_source_range = resource_block.source_range
+
+            if on_macos_blocks.length > 1
+              problem "there can only be one `on_macos` block in a resource block."
+              next
+            end
+
+            if on_linux_blocks.length > 1
+              problem "there can only be one `on_linux` block in a resource block."
+              next
+            end
+
+            if on_macos_blocks.length == 1 && on_linux_blocks.length.zero?
+              problem "you need to define an `on_linux` block within your resource block."
+              next
+            end
+
+            if on_macos_blocks.length.zero? && on_linux_blocks.length == 1
+              problem "you need to define an `on_macos` block within your resource block."
+              next
+            end
+
+            on_macos_block = on_macos_blocks.first
+            on_linux_block = on_linux_blocks.first
+
+            child_nodes = on_macos_block.body.child_nodes
+            if child_nodes[0].method_name.to_s != "url" && child_nodes[1].method_name.to_s != "sha256"
+              problem "only an url and a sha256 (in the right order) are allowed in a `on_macos` " \
+                "block within a resource block."
+              next
+            end
 
-              @offensive_nodes = check_precedence(preceding_component, succeeding_component)
-              component_problem @offensive_nodes[0], @offensive_nodes[1] if @offensive_nodes
+            child_nodes = on_linux_block.body.child_nodes
+            if child_nodes[0].method_name.to_s != "url" && child_nodes[1].method_name.to_s != "sha256"
+              problem "only an url and a sha256 (in the right order) are allowed in a `on_linux` " \
+                "block within a resource block."
             end
           end
         end
 
-        # `aspell`: options and resources should be grouped by language
-        WHITELIST = %w[
-          aspell
-        ].freeze
+        def check_on_os_block_content(component_precedence_list, on_os_block)
+          _, offensive_node = check_order(component_precedence_list, on_os_block.body)
+          component_problem(*offensive_node) if offensive_node
+          on_os_block.body.child_nodes.each do |child|
+            valid_node = depends_on_node?(child)
+            next unless child.instance_of? RuboCop::AST::SendNode
```

Comment:
This warrants a comment explaining why this class is different

Reply:
I added a comment. Rubocop is a little bit weird and we probably mis-used it in some places. We have been using regexes (like in the depends_on_node function), but it's probably better to check for instances of classes and use the right methods, like I did with `method_name` below. We are mixing both strategies here, I could probably rewrite this with regexes as done before, but I think it is more fragile to look at the string representation of the nodes and try to use regexes on them ...